src/transformers/generation/logits_process.py

```diff
@@ -355,17 +355,55 @@ def __init__(self, penalty: float, prompt_ignore_length: Optional[int] = None):
 
         self.penalty = penalty
         self.prompt_ignore_length = prompt_ignore_length
+        self.logits_indices = None
+        self.cumulative_seqlens_q = None
+
+    def set_continuous_batching_context(self, logits_indices: torch.Tensor, cumulative_seqlens_q: torch.Tensor):
+        self.logits_indices = logits_indices
+        self.cumulative_seqlens_q = cumulative_seqlens_q
 
     @add_start_docstrings(LOGITS_PROCESSOR_INPUTS_DOCSTRING)
     def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
         if self.prompt_ignore_length:
             input_ids = input_ids[:, self.prompt_ignore_length :]
 
-        score = torch.gather(scores, 1, input_ids)
+        input_ids = input_ids.to(torch.int64)
+
+        if scores.dim() == 3:
+            if self.logits_indices is not None and self.cumulative_seqlens_q is not None:
+                batch_size, seq_len, vocab_size = scores.shape
+                last_positions = self.logits_indices
+                last_scores = scores[0, last_positions, :]
+                token_mask = torch.zeros_like(last_scores, dtype=torch.bool)
+                cu_seq_lens = self.cumulative_seqlens_q
+                for i in range(len(cu_seq_lens) - 1):
+                    start_idx = cu_seq_lens[i]
+                    end_idx = cu_seq_lens[i + 1]
+                    sequence_tokens = input_ids[start_idx:end_idx]
+                    token_mask[i].scatter_(0, sequence_tokens, True)
+                penalty_scores = torch.where(last_scores < 0, last_scores * self.penalty, last_scores / self.penalty)
+                scores[0, last_positions, :] = torch.where(token_mask, penalty_scores, last_scores)
+            else:
+                batch_size, seq_len, vocab_size = scores.shape
+                last_scores = scores[:, -1, :]
+                token_mask = torch.zeros_like(last_scores, dtype=torch.bool)
+                if input_ids.dim() == 1:
+                    unique_tokens = torch.unique(input_ids)
+                    token_mask.scatter_(1, unique_tokens.unsqueeze(0), True)
```

Comment:
```suggestion
    batch_size, seq_len, vocab_size = scores.shape
    last_positions = self.logits_indices
    last_scores = scores[0, last_positions, :]
    
    # Prepare token mask
    token_mask = torch.zeros_like(last_scores, dtype=torch.bool)
    cu_seq_lens = self.cumulative_seqlens_q
    lengths = cu_seq_lens[1:] - cu_seq_lens[:-1]
    seq_indices = torch.repeat_interleave(torch.arange(len(lengths), device=input_ids.device), lengths)
    token_mask[seq_indices, input_ids] = True
    
    # Apply penalty
    penalty_scores = torch.where(last_scores < 0, last_scores * self.penalty, last_scores / self.penalty)
    scores[0, last_positions, :] = torch.where(token_mask, penalty_scores, last_scores)
```
I think we can vectorize !